File: backend/product-service/src/main/java/com/shopflow/productservice/service/ProductService.java
package com.shopflow.productservice.service;

import com.shopflow.productservice.dto.ProductRequest;
import com.shopflow.productservice.dto.ProductResponse;
import com.shopflow.productservice.entity.Product;
import com.shopflow.productservice.exception.ResourceNotFoundException;
import com.shopflow.productservice.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;

    public ProductResponse addProduct(ProductRequest request) {

        Product product = Product.builder()
                .name(request.getName())
                .description(request.getDescription())
                .price(request.getPrice())
                .category(request.getCategory())
                .imageUrl(request.getImageUrl())
                .build();

        Product savedProduct = productRepository.save(product);

        return mapToResponse(savedProduct);
    }

    public List<Product> getAllProducts() {

        return productRepository.findAll();

    }

    public Product getProductById(Long id) {

        return productRepository.findById(id)
                .orElseThrow(
                        () -> new ResourceNotFoundException(
                                "Product not found with id : " + id
                        )
                );

    }

    public Product updateProduct(
            Long id,
            ProductRequest request
    ) {

        Product product = productRepository.findById(id)
                .orElseThrow(
                        () -> new ResourceNotFoundException(
                                "Product not found with id : " + id
                        )
                );

        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setCategory(request.getCategory());
        product.setImageUrl(request.getImageUrl());

        return productRepository.save(product);

    }

    public void deleteProduct(Long id) {

        if (!productRepository.existsById(id)) {

            throw new ResourceNotFoundException(
                    "Product not found with id : " + id
            );

        }

        productRepository.deleteById(id);

    }

    public List<Product> searchProducts(String keyword) {

        return productRepository.findByNameContainingIgnoreCase(keyword);

    }

    public List<Product> getProductsByCategory(String category) {

        return productRepository.findByCategoryIgnoreCase(category);

    }

    private ProductResponse mapToResponse(Product product) {

        return ProductResponse.builder()
                .id(product.getId())
                .name(product.getName())
                .description(product.getDescription())
                .price(product.getPrice())
                .category(product.getCategory())
                .imageUrl(product.getImageUrl())
                .build();

    }

}

File: frontend/src/services/productService.js
import api from "../api/axios";

export async function getAllProducts() {

  const response = await api.get(
    "/api/products"
  );

  return response.data;

}

export async function getProductById(id) {

  const response = await api.get(
    `/api/products/${id}`
  );

  return response.data;

}

export async function addProduct(product) {

  const response = await api.post(
    "/api/products",
    product
  );

  return response.data;

}

export async function updateProduct(id, product) {

  const response = await api.put(
    `/api/products/${id}`,
    product
  );

  return response.data;

}

export async function deleteProduct(id) {

  const response = await api.delete(
    `/api/products/${id}`
  );

  return response.data;

}

export async function getProductsByCategory(category) {

  const response = await api.get(
    `/api/products/category/${category}`
  );

  return response.data;

}

export async function searchProducts(keyword) {

  const response = await api.get(
    `/api/products/search/${keyword}`
  );

  return response.data;

}